File: src/main/ruby/diffj/ast/ctor.rb
#!/usr/bin/jruby -w
# -*- ruby -*-

require 'rubygems'
require 'riel'
require 'java'
require 'diffj/ast/item'
require 'diffj/ast/function'

include Java

module DiffJ
  class CtorComparator < FunctionComparator
    include Loggable

    def compare_parameters from, to
      super from.parameters, to.parameters
    end

    def compare_throws from, to
      super from, from.throws_list, to, to.throws_list
    end
    
    def get_code ctor
      # removes all tokens up to the first left brace. This is because ctors
      # don't have their own blocks, unlike methods.
        
      code = ctor.get_child_tokens
      first_lbrace = code.index { |tk| tk.kind == ::Java::net.sourceforge.pmd.ast.JavaParserConstants::LBRACE }

      code[first_lbrace .. -1]
    end

    def compare_bodies from, to
      from_code = get_code from
      to_code = get_code to
        
      compare_code from.fullname, from_code, to.fullname, to_code
    end

    def compare from, to
      info "from: #{from}"
      info "to  : #{to}"
      
      compare_parameters from, to
      compare_throws from, to
      compare_bodies from, to
    end
  end
end

File: src/main/java/org/incava/diffj/Ctor.java
package org.incava.diffj;

import java.util.Iterator;
import java.util.List;
import net.sourceforge.pmd.ast.ASTConstructorDeclaration;
import net.sourceforge.pmd.ast.ASTFormalParameters;
import net.sourceforge.pmd.ast.ASTNameList;
import net.sourceforge.pmd.ast.JavaParserConstants;
import net.sourceforge.pmd.ast.SimpleNode;
import net.sourceforge.pmd.ast.Token;
import org.incava.pmdx.CtorUtil;
import org.incava.pmdx.SimpleNodeUtil;

public class Ctor extends Function {
    private final ASTConstructorDeclaration ctor;

    public Ctor(ASTConstructorDeclaration ctor) {
        super(ctor);
        this.ctor = ctor;
    }

    public void diff(Ctor toCtor, Differences differences) {
        compareAccess(toCtor, differences);
        compareParameters(toCtor, differences);
        compareThrows(toCtor, differences);
        compareBodies(toCtor, differences);
    }

    protected ASTFormalParameters getParameters() {
        return CtorUtil.getParameters(ctor);
    }

    protected ASTNameList getThrowsList() {
        return CtorUtil.getThrowsList(ctor);
    }

    protected String getName() {
        return CtorUtil.getFullName(ctor);
    }
    
    protected void compareParameters(Ctor toCtor, Differences differences) {
        ASTFormalParameters fromParam = getParameters();
        ASTFormalParameters toParams = toCtor.getParameters();
        compareParameters(fromParam, toParams, differences);
    }

    protected List<Token> getCodeTokens() {
        // removes all tokens up to the first left brace. This is because ctors
        // don't have their own blocks, unlike methods.
        
        List<Token> children = SimpleNodeUtil.getChildTokens(ctor);
        
        Iterator<Token> it = children.iterator();
        while (it.hasNext()) {
            Token tk = it.next();
            if (tk.kind == JavaParserConstants.LBRACE) {
                break;
            }
            else {
                it.remove();
            }
        }

        return children;
    }

    protected void compareBodies(Ctor toCtor, Differences differences) {
        List<Token> fromCode = getCodeTokens();
        List<Token> toCode = toCtor.getCodeTokens();
        String fromName = getName();
        compareCode(fromName, fromCode, toCode, differences);
    }
}
